### PR TITLE
Make helper methods of `WriterBasedJsonGenerator` non-final to allow overriding

### PR DESCRIPTION
**Pull request by zhangOranges**
The protected method removes the final to allow users to rewrite the method and implement their own logic

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/core/json/WriterBasedJsonGenerator.java [hunk 1]
```diff
@@ -17,10 +17,10 @@
 public class WriterBasedJsonGenerator
     extends JsonGeneratorImpl
 {
-    protected final static int SHORT_WRITE = 32;
+    protected static int SHORT_WRITE = 32;
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/json/WriterBasedJsonGenerator.java:20]: No, let's leave static fields final.

File: src/main/java/com/fasterxml/jackson/core/json/WriterBasedJsonGenerator.java [hunk 2]
```diff
@@ -17,10 +17,10 @@
...
 
-    protected final static char[] HEX_CHARS_UPPER = CharTypes.copyHexChars(true);
-    protected final static char[] HEX_CHARS_LOWER = CharTypes.copyHexChars(false);
+    protected static char[] HEX_CHARS_UPPER = CharTypes.copyHexChars(true);
+    protected static char[] HEX_CHARS_LOWER = CharTypes.copyHexChars(false);
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/json/WriterBasedJsonGenerator.java:23]: Same here, need to remain final.

**Comment by cowtowncoder** [comment]
I am ok with targeted removal of `final` (with all the caveats that sub-classing of this type is strongly discouraged as these methods are not part of API in the sense that no backwards-compatibility is implied across minor versions) for methods. But static fields should remain final, they are not meant to be changed.

**Comment by cowtowncoder** [comment]
Aside from requested change wrt static fields, one thing I'd need before merging is CLA (unless already received one), from:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

The usual way is to print it, fill & sign, scan/photo, email to `cla` at fasterxml dot com.
This is needed once before the first contribution; after that good for all future PRs.

**Comment by zhangOranges** [comment]
Okay, I want to add some of my own logical operations before writing to outBuffer, but I couldn't find the provided entry, so I saw the WriterBasedJsonGenerator class. However, some of the protected methods were declared final, which confused me. Is there any other better way for me to operate, similar to an AOP

**Comment by cowtowncoder** [comment]
@zhangOranges One thing that might work `JsonGeneratorDelegate`:

    src/main/java/com/fasterxml/jackson/core/util/JsonGeneratorDelegate.java

which could intercept calls. It depends on kinds of changes you need to make.
You basically use it to "wrap" a real generator, and by default it just delegates all calls; but you can override whatever public methods you want. Since it operates on public API it is more limited, but also much safer and easier to maintain approach.

**Comment by zhangOranges** [comment]
> JsonGeneratorDelegate

{
 //test
 "demo":"1"
}   Annotation test is what I want to add, JsonGeneratorDelegate may not meet my needs, right？

**Comment by zhangOranges** [comment]
> @zhangOranges One thing that might work `JsonGeneratorDelegate`:
> 
> ```
> src/main/java/com/fasterxml/jackson/core/util/JsonGeneratorDelegate.java
> ```
> 
> which could intercept calls. It depends on kinds of changes you need to make. You basically use it to "wrap" a real generator, and by default it just delegates all calls; but you can override whatever public methods you want. Since it operates on public API it is more limited, but also much safer and easier to maintain approach.

I want to add some other things before writing the correct JSON key value

**Comment by zhangOranges** [comment]
> Aside from requested change wrt static fields, one thing I'd need before merging is CLA (unless already received one), from:
> 
> https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf
> 
> The usual way is to print it, fill & sign, scan/photo, email to `cla` at fasterxml dot com. This is needed once before the first contribution; after that good for all future PRs.

Is there any other simple signature method? I currently don't have a printer

**Comment by cowtowncoder** [comment]
> > @zhangOranges One thing that might work `JsonGeneratorDelegate`:
> > ```
> > src/main/java/com/fasterxml/jackson/core/util/JsonGeneratorDelegate.java
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which could intercept calls. It depends on kinds of changes you need to make. You basically use it to "wrap" a real generator, and by default it just delegates all calls; but you can override whatever public methods you want. Since it operates on public API it is more limited, but also much safer and easier to maintain approach.
> 
> I want to add some other things before writing the correct JSON key value

It would be possible to override `writeFieldName()`, call `writeRaw()` on delegate before forwarding `writeFieldName()`.

**Comment by cowtowncoder** [comment]
> > Aside from requested change wrt static fields, one thing I'd need before merging is CLA (unless already received one), from:
> > https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf
> > The usual way is to print it, fill & sign, scan/photo, email to `cla` at fasterxml dot com. This is needed once before the first contribution; after that good for all future PRs.
> 
> Is there any other simple signature method? I currently don't have a printer

What some contributors have done is to use an app that takes PDF, allows adding text to fill in the fields (add text in approximate location). That works for me.

**Comment by zhangOranges** [comment]
> > > Aside from requested change wrt static fields, one thing I'd need before merging is CLA (unless already received one), from:
> > > https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf
> > > The usual way is to print it, fill & sign, scan/photo, email to `cla` at fasterxml dot com. This is needed once before the first contribution; after that good for all future PRs.
> > 
> > 
> > Is there any other simple signature method? I currently don't have a printer
> 
> What some contributors have done is to use an app that takes PDF, allows adding text to fill in the fields (add text in approximate location). That works for me.

Which email to send after signing

**Comment by zhangOranges** [comment]
> > > @zhangOranges One thing that might work `JsonGeneratorDelegate`:
> > > ```
> > > src/main/java/com/fasterxml/jackson/core/util/JsonGeneratorDelegate.java
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > which could intercept calls. It depends on kinds of changes you need to make. You basically use it to "wrap" a real generator, and by default it just delegates all calls; but you can override whatever public methods you want. Since it operates on public API it is more limited, but also much safer and easier to maintain approach.
> > 
> > 
> > I want to add some other things before writing the correct JSON key value
> 
> It would be possible to override `writeFieldName()`, call `writeRaw()` on delegate before forwarding `writeFieldName()`.

WriteFieldName will add a comma to the end of the previous execution, so this method is not feasible

**Comment by cowtowncoder** [comment]
> > > > Aside from requested change wrt static fields, one thing I'd need before merging is CLA (unless already received one), from:
> > > > https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf
> > > > The usual way is to print it, fill & sign, scan/photo, email to `cla` at fasterxml dot com. This is needed once before the first contribution; after that good for all future PRs.
> > > 
> > > 
> > > Is there any other simple signature method? I currently don't have a printer
> > 
> > 
> > What some contributors have done is to use an app that takes PDF, allows adding text to fill in the fields (add text in approximate location). That works for me.
> 
> Which email to send after signing

I mentioned this earlier; "cla@fasterxml.com"

**Comment by zhangOranges** [comment]
> > > > > Aside from requested change wrt static fields, one thing I'd need before merging is CLA (unless already received one), from:
> > > > > https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf
> > > > > The usual way is to print it, fill & sign, scan/photo, email to `cla` at fasterxml dot com. This is needed once before the first contribution; after that good for all future PRs.
> > > > 
> > > > 
> > > > Is there any other simple signature method? I currently don't have a printer
> > > 
> > > 
> > > What some contributors have done is to use an app that takes PDF, allows adding text to fill in the fields (add text in approximate location). That works for me.
> > 
> > 
> > Which email to send after signing
> 
> I mentioned this earlier; "[cla@fasterxml.com](mailto:cla@fasterxml.com)"

ok

**Comment by zhangOranges** [comment]
> > > > > Aside from requested change wrt static fields, one thing I'd need before merging is CLA (unless already received one), from:
> > > > > https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf
> > > > > The usual way is to print it, fill & sign, scan/photo, email to `cla` at fasterxml dot com. This is needed once before the first contribution; after that good for all future PRs.
> > > > 
> > > > 
> > > > Is there any other simple signature method? I currently don't have a printer
> > > 
> > > 
> > > What some contributors have done is to use an app that takes PDF, allows adding text to fill in the fields (add text in approximate location). That works for me.
> > 
> > 
> > Which email to send after signing
> 
> I mentioned this earlier; "[cla@fasterxml.com](mailto:cla@fasterxml.com)"

has been sent

**Comment by cowtowncoder** [comment]
Ok received the first CLA and while it's almost good asked for minor edit (so there's name and github id both).

**Comment by zhangOranges** [comment]
orange
***@***.***



&nbsp;





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "FasterXML/jackson-core"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年6月14日(星期五) 凌晨5:04
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [FasterXML/jackson-core] Update WriterBasedJsonGenerator.java (PR #1305)





 
Ok received the first CLA and while it's almost good asked for minor edit (so there's name and github id both).

**Comment by zhangOranges** [comment]
> Ok received the first CLA and while it's almost good asked for minor edit (so there's name and github id both).

has been sent

**Comment by cowtowncoder** [comment]
CLA received.

**Review by cowtowncoder — APPROVED** [review]

**Pull request merged**